File: src/cpu.rs
use std::{hint::unreachable_unchecked, ops::{Index, IndexMut}, u64, ops::{Sub, Add, BitAnd, BitXor}};

use crate::isa::{Cond, Op, Register, Y86, Y86BLANK};

#[derive(Default)]
pub struct Flags {
    sf: bool,
    zf: bool,
    of: bool,
}

#[derive(Default)]
pub struct RegFile {
    rax: u64,
    rcx: u64,
    rdx: u64,
    rbx: u64,
    rsp: u64,
    rbp: u64,
    rsi: u64,
    rdi: u64,
    r8: u64,
    r9: u64,
    r10: u64,
    r11: u64,
    r12: u64,
    r13: u64,
    r14: u64,
}

impl Index<Register> for RegFile {
    type Output = u64;
    fn index(&self, index: Register) -> &Self::Output {
        match index {
            Register::RAX => &self.rax,
            Register::RCX => &self.rcx,
            Register::RDX => &self.rdx,
            Register::RBX => &self.rbx,
            Register::RSP => &self.rsp,
            Register::RBP => &self.rbp,
            Register::RSI => &self.rsi,
            Register::RDI => &self.rdi,
            Register::R8 => &self.r8,
            Register::R9 => &self.r9,
            Register::R10 => &self.r10,
            Register::R11 => &self.r11,
            Register::R12 => &self.r12,
            Register::R13 => &self.r13,
            Register::R14 => &self.r14,
        }
    }
}

impl IndexMut<Register> for RegFile {
    fn index_mut(&mut self, index: Register) -> &mut Self::Output {
        match index {
            Register::RAX => &mut self.rax,
            Register::RCX => &mut self.rcx,
            Register::RDX => &mut self.rdx,
            Register::RBX => &mut self.rbx,
            Register::RSP => &mut self.rsp,
            Register::RBP => &mut self.rbp,
            Register::RSI => &mut self.rsi,
            Register::RDI => &mut self.rdi,
            Register::R8 => &mut self.r8,
            Register::R9 => &mut self.r9,
            Register::R10 => &mut self.r10,
            Register::R11 => &mut self.r11,
            Register::R12 => &mut self.r12,
            Register::R13 => &mut self.r13,
            Register::R14 => &mut self.r14,
        }
    }
}

#[allow(non_snake_case)]
pub struct CPU {
    flags: Flags,
    registers: RegFile,
    memory: [u8; 0x400],
    program_counter: usize,
    // begin ifun
    condition: Cond,
    cnd: bool,
    op: Op,
    // end ifun
    curr: Y86BLANK,
    rA: Register,
    rB: Register,
    valA: u64,
    valB: u64,
    valC: u64,
    valE: u64,
    valM: u64,
    valP: usize,
    pub(super) stat: bool,
}

impl Default for CPU {
    fn default() -> Self {
        CPU {
            memory: [0; 0x400],
            flags: Default::default(),
            registers: Default::default(),
            program_counter: Default::default(),
            condition: Default::default(),
            cnd: Default::default(),
            op: Default::default(),
            curr: Default::default(),
            rA: Default::default(),
            rB: Default::default(),
            valA: Default::default(),
            valB: Default::default(),
            valC: Default::default(),
            valE: Default::default(),
            valM: Default::default(),
            valP: Default::default(),
            stat: Default::default(),
        }
    }
}

pub trait FeDeExMemWBPC {
    fn fetch(&mut self);
    fn decode(&mut self);
    fn execute(&mut self);
    fn memory(&mut self);
    fn writeback(&mut self);
    fn program_counter(&mut self);
}

impl FeDeExMemWBPC for CPU {
    fn fetch(&mut self) {
        let first = self.memory[self.program_counter];
        match first & 0xf0 {
            0x00 => {
                self.valP = self.program_counter + 1;
                self.curr = Y86BLANK::HALT;
            },
            0x10 => {
                self.valP = self.program_counter + 1;
                self.curr = Y86BLANK::NOOP;
            },
            0x20 => { 
                self.condition = (first & 0x0f).into();
                let regs = self.memory[self.program_counter + 1];
                self.rA = ((regs & 0xf0) >> 4).into();
                self.rB = (regs & 0x0f).into();

                self.valP = self.program_counter + 2;
                self.curr = Y86BLANK::CMOV;
            }
            0x30 => {
                let regs = self.memory[self.program_counter + 1];
                self.rB = (regs & 0x0f).into();

                let mut res: u64 = 0;
                for i in 0..8 {
                    res += (self.memory[self.program_counter + 2 + i] << (8 * i)) as u64;
                }

                self.valC = res;

                self.valP = self.program_counter + 10;
                self.curr = Y86BLANK::IRMOVQ;
            }
            0x40 => {
                let regs = self.memory[self.program_counter + 1];
                self.rA = ((regs & 0xf0) >> 4).into();
                self.rB = (regs & 0x0f).into();

                let mut res: u64 = 0;
                for i in 0..8 {
                    res += (self.memory[self.program_counter + 2 + i] << (8 * i)) as u64;
                }

                self.valC = res;

                self.valP = self.program_counter + 10;
                self.curr = Y86BLANK::RMMOVQ;
            }
            0x50 => {
                let regs = self.memory[self.program_counter + 1];
                self.rA = ((regs & 0xf0) >> 4).into();
                self.rB = (regs & 0x0f).into();

                let mut res: u64 = 0;
                for i in 0..8 {
                    res += (self.memory[self.program_counter + 2 + i] << (8 * i)) as u64;
                }

                self.valC = res;

                self.valP = self.program_counter + 10;
                self.curr = Y86BLANK::MRMOVQ;
            }
            0x60 => { 
                self.op = (first & 0x0f).into();

                let regs = self.memory[self.program_counter + 1];
                self.rA = ((regs & 0xf0) >> 4).into();
                self.rB = (regs & 0x0f).into();

                self.curr = Y86BLANK::OPQ;
            }
            0x70 => {
                self.condition = (first & 0x0f).into();

                let mut res: u64 = 0;
                for i in 0..8 {
                    res += (self.memory[self.program_counter + 1 + i] << (8 * i)) as u64;
                }
                self.valC = res;

                self.valP = self.program_counter + 9;
                self.curr = Y86BLANK::J;
            }
            0x80 => {
                let mut res: u64 = 0;
                for i in 0..8 {
                    res += (self.memory[self.program_counter + 1 + i] << (8 * i)) as u64;
                }
                self.valC = res;

                self.valP = self.program_counter + 9;
                self.curr = Y86BLANK::CALL;
            }
            0x90 => {
                self.valP = self.program_counter + 1;
                self.curr = Y86BLANK::RET;
            },
            0xa0 => {
                let regs = self.memory[self.program_counter + 1];
                self.rA = ((regs & 0xf0) >> 4).into();
                
                self.valP = self.program_counter + 2;
                self.curr = Y86BLANK::PUSHQ;
            }
            0xb0 => { 
                let regs = self.memory[self.program_counter + 1];
                self.rA = ((regs & 0xf0) >> 4).into();
                
                self.valP = self.program_counter + 2;
                self.curr = Y86BLANK::POPQ;
            }
            _ => unsafe { unreachable_unchecked() }
        }
    }

    fn decode(&mut self) {
        match &self.curr {
            Y86BLANK::CMOV => {
                self.valA = self.registers[self.rA];
            }
            Y86BLANK::RMMOVQ => {
                self.valA = self.registers[self.rA];
                self.valB = self.registers[self.rB];
            }
            Y86BLANK::MRMOVQ => {
                self.valB = self.registers[self.rB];
            }
            Y86BLANK::OPQ => {
                self.valA = self.registers[self.rA];
                self.valB = self.registers[self.rB];
            }
            Y86BLANK::CALL => {
                self.valB = self.registers[Register::RSP];
            }
            Y86BLANK::RET => {
                self.valA = self.registers[Register::RSP];
                self.valB = self.registers[Register::RSP];
            }
            Y86BLANK::PUSHQ => {
                self.valA = self.registers[self.rA];
                self.valB = self.registers[Register::RSP];
            }
            Y86BLANK::POPQ => {
                self.valA = self.registers[Register::RSP];
                self.valB = self.registers[Register::RSP];
            }
            _ => (),
        }
    }

    fn execute(&mut self) {
        match &self.curr {
            Y86BLANK::HALT => self.stat = false,
            // Condition handled in fetch
            Y86BLANK::CMOV => self.valE = self.valA,
            Y86BLANK::IRMOVQ => self.valE = self.valC,
            Y86BLANK::RMMOVQ | Y86BLANK::MRMOVQ => self.valE = self.valB + self.valC,
            Y86BLANK::OPQ => { 
                self.valE = [u64::wrapping_add, u64::wrapping_sub, u64::bitand, u64::bitxor][self.op as usize](self.valA, self.valB);
                match self.op {
                    Op::Add => self.flags.of = self.valE < self.valA || self.valE < self.valA,
                    Op::Sub => self.flags.of = self.valE > self.valA || self.valE > self.valA,
                    _ => (),
                }
                if self.valE == 0 {
                    self.flags.zf = true;
                }
                if self.valE & 1 << 63 != 0 {
                    self.flags.sf = true;
                }
            }
            Y86BLANK::J => {
                let (sf, zf, of) = (self.flags.sf, self.flags.zf, self.flags.of);
                self.cnd = match self.condition {
                    Cond::None => true,
                    Cond::Le   => sf || zf,
                    Cond::L    => sf && zf,
                    Cond::E    => zf,
                    Cond::Ne   => !zf,
                    Cond::Ge   => !sf || zf,
                    Cond::G    => !sf && !zf,
                };
            }
            // Condition handled in fetch for [`Y86BLANK::J`]
            Y86BLANK::CALL => self.valE = self.valB - 8,
            Y86BLANK::RET => self.valE = self.valB + 8,
            Y86BLANK::PUSHQ => self.valE = self.valB - 8,
            Y86BLANK::POPQ => self.valE = self.valB + 8,
            _ => (),
        }
    }

    fn memory(&mut self) {
        match &self.curr {
            Y86BLANK::RMMOVQ => { 
                for i in 0..8 {
                    self.memory[self.valE as usize + i] = ((self.valA >> (8 * i)) & 0xff) as u8;
                }
            }
            Y86BLANK::MRMOVQ => {
                let mut res = 0;
                for i in 0..8 {
                    res += (self.memory[self.valE as usize + i] << (8 * i)) as u64;
                }
                self.valM = res;
            }
            Y86BLANK::CALL => {
                for i in 0..8 {
                    self.memory[self.valE as usize + i] = ((self.valP >> (8 * i)) & 0xff) as u8;
                }
            }
            Y86BLANK::RET => {
                let mut res = 0;
                for i in 0..8 {
                    res += (self.memory[self.valA as usize + i] << (8 * i)) as u64;
                }
                self.valM = res;
            }
            Y86BLANK::PUSHQ => {
                for i in 0..8 {
                    self.memory[self.valE as usize + i] = ((self.valA >> (8 * i)) & 0xff) as u8;
                }
            }
            Y86BLANK::POPQ => {
                let mut res = 0;
                for i in 0..8 {
                    res += (self.memory[self.valA as usize + i] << (8 * i)) as u64;
                }
                self.valM = res;
            }
            _ => (),
        }
    }

    fn writeback(&mut self) {
        match &self.curr {
            Y86BLANK::CMOV   => if self.cnd { self.registers[self.rB] = self.valE },
            Y86BLANK::IRMOVQ => self.registers[self.rB] = self.valE,
            Y86BLANK::MRMOVQ => self.registers[self.rA] = self.valM,
            Y86BLANK::OPQ    => self.registers[self.rB] = self.valE,

            Y86BLANK::CALL   => self.registers[Register::RSP] = self.valE,
            Y86BLANK::RET    => self.registers[Register::RSP] = self.valE,
            Y86BLANK::PUSHQ  => self.registers[Register::RSP] = self.valE,
            Y86BLANK::POPQ   => {
                self.registers[Register::RSP] = self.valE;
                self.registers[self.rA] = self.valM;
            },
            _ => (),
        }
    }

    fn program_counter(&mut self) {
        match &self.curr {
            Y86BLANK::HALT   => self.program_counter = 0,
            Y86BLANK::NOOP   => self.program_counter = self.valP,
            Y86BLANK::CMOV   => self.program_counter = self.valP,
            Y86BLANK::IRMOVQ => self.program_counter = self.valP,
            Y86BLANK::RMMOVQ => self.program_counter = self.valP,
            Y86BLANK::MRMOVQ => self.program_counter = self.valP,
            Y86BLANK::OPQ    => self.program_counter = self.valP,
            Y86BLANK::J      => self.program_counter = if (self.cnd) { self.valC as usize } else { self.valP },
            Y86BLANK::CALL   => self.program_counter = self.valC as usize,
            Y86BLANK::RET    => self.program_counter = self.valM as usize,
            Y86BLANK::PUSHQ  => self.program_counter = self.valP,
            Y86BLANK::POPQ   => self.program_counter = self.valP,
        }
    }
}

File: cy86/cpu.c
#include "cpu.h"

CPU default_cpu() {
    CPU cpu = {
        .flags = {
            .zf = false,
            .sf = false,
            .of = false,
        },
        .registers = {
            .rax = 0,
            .rcx = 0,
            .rdx = 0,
            .rbx = 0,
            .rsp = 0,
            .rbp = 0,
            .rsi = 0,
            .rdi = 0,
            .r8 = 0,
            .r9 = 0,
            .r10 = 0,
            .r11 = 0,
            .r12 = 0,
            .r13 = 0,
            .r14 = 0,
        },
        .memory = { 0 },
        .program_counter = 0,
        .condition = None,
        .cnd = false,
        .op = 0,
        .curr = 0,
        .rA   = 0,
        .rB   = 0,
        .valA = 0,
        .valB = 0,
        .valC = 0,
        .valE = 0,
        .valM = 0,
        .valP = 0,
        .stat = false,
    };
    return cpu;
}

u64 get_reg(RegFile *file, Register index) {
    return ((u64*)file)[index];
}

void set_reg(RegFile *file, Register index, u64 value) {
    ((u64*)file)[index] = value;
}

void fetch(CPU* cpu) {
    u8 op = cpu->memory[cpu->program_counter];
    u8 regs;
    u64 res;
    switch (op & 0xf0) {
        case 0x00:
            cpu->curr = HALT;
            cpu->valP = cpu->program_counter + 1;
            break;
        case 0x10:
            cpu->curr = NOOP;
            cpu->valP = cpu->program_counter + 1;
            break;
        case 0x20:
            cpu->curr = CMOV;
            cpu->condition = (op & 0x0f);
            regs = cpu->memory[cpu->program_counter + 1];
            cpu->rA = (regs & 0xf0) >> 4;
            cpu->rB = (regs & 0xf0);
            cpu->valP = cpu->program_counter + 2;
            break;
        case 0x30:
            cpu->curr = IRMOVQ;
            regs = cpu->memory[cpu->program_counter + 1];
            cpu->rB = (regs & 0xf0);

            res = *((u64*)(cpu->memory + cpu->program_counter + 2));
            cpu->valC = res;
            cpu->valP = cpu->program_counter + 10;
            break;
        case 0x40:
            cpu->curr = RMMOVQ;
            regs = cpu->memory[cpu->program_counter + 1];
            cpu->rA = (regs & 0xf0) >> 4;
            cpu->rB = (regs & 0xf0);
            res = *((u64*)(cpu->memory + cpu->program_counter + 2));
            cpu->valP = cpu->program_counter + 10;
            break;
        case 0x50:
            cpu->curr = MRMOVQ;
            regs = cpu->memory[cpu->program_counter + 1];
            cpu->rA = (regs & 0xf0) >> 4;
            cpu->rB = (regs & 0xf0);
            res = *((u64*)(cpu->memory + cpu->program_counter + 2));
            cpu->valP = cpu->program_counter + 10;
            break;
        case 0x60:
            cpu->curr = OPQ;
            cpu->op = op & 0x0f;
            regs = cpu->memory[cpu->program_counter + 1];
            cpu->rA = (regs & 0xf0) >> 4;
            cpu->rB = (regs & 0xf0);
            cpu->valP = cpu->program_counter + 2;
            break;
        case 0x70:
            cpu->curr = J;
            cpu->condition = (op & 0x0f);
            cpu->valC = *((u64*)(cpu->memory + cpu->program_counter + 1));
            cpu->valP = cpu->program_counter + 9;
            break;
        case 0x80:
            cpu->curr = CALL;
            cpu->valC = *((u64*)(cpu->memory + cpu->program_counter + 1));
            cpu->valP = cpu->program_counter + 9;
            break;
        case 0x90:
            cpu->curr = RET;
            cpu->valP = cpu->program_counter + 1;
            break;
        case 0xa0:
            cpu->curr = PUSHQ;
            regs = cpu->memory[cpu->program_counter + 1];
            cpu->rA = (regs & 0xf0) >> 4;
            cpu->rB = (regs & 0xf0);
            cpu->valP = cpu->program_counter + 2;
            break;
        case 0xb0:
            cpu->curr = POPQ;
            regs = cpu->memory[cpu->program_counter + 1];
            cpu->rA = (regs & 0xf0) >> 4;
            cpu->rB = (regs & 0xf0);
            cpu->valP = cpu->program_counter + 2;
            break;
    }
}

void decode(CPU* cpu) {
    switch (cpu->curr) {
        case CMOV:
            cpu->valA = get_reg(&cpu->registers, cpu->rA);
            break;
        case RMMOVQ:
            cpu->valA = get_reg(&cpu->registers, cpu->rA);
            cpu->valB = get_reg(&cpu->registers, cpu->rB);
            break;
        case MRMOVQ:
            cpu->valB = get_reg(&cpu->registers, cpu->rB);
            break;
        case OPQ:
            cpu->valA = get_reg(&cpu->registers, cpu->rA);
            cpu->valB = get_reg(&cpu->registers, cpu->rB);
            break;
        case CALL:
            cpu->valB = get_reg(&cpu->registers, RSP);
            break;
        case RET:
            cpu->valA = get_reg(&cpu->registers, RSP);
            cpu->valB = get_reg(&cpu->registers, RSP);
            break;
        case PUSHQ:
            cpu->valA = get_reg(&cpu->registers, cpu->rA);
            cpu->valB = get_reg(&cpu->registers, RSP);
            break;
        case POPQ:
            cpu->valA = get_reg(&cpu->registers, RSP);
            cpu->valB = get_reg(&cpu->registers, RSP);
        default:
            break;
    }
}

u64 add(u64 a, u64 b) {
    return a + b;
}

u64 sub(u64 a, u64 b) {
    return a - b;
}

u64 and(u64 a, u64 b) {
    return a & b;
}

u64 xor(u64 a, u64 b) {
    return a ^ b;
}

void execute(CPU* cpu) {
    u64 (*actions[4])(u64, u64) = {add, sub, and, xor};
    bool sf;
    bool zf;
    bool of;
    switch (cpu->curr) {
        case HALT:
            cpu->stat = false;
            break;
        case CMOV:
            cpu->valE = cpu->valA;
            break;
        case IRMOVQ:
            cpu->valE = cpu->valC;
            break;
        case RMMOVQ:
        case MRMOVQ:
            cpu->valE = cpu->valB + cpu->valC;
            break;
        case OPQ:
            cpu->valE = actions[cpu->op](cpu->valA, cpu->valB);
            if (cpu->op == Add)
                cpu->flags.of = cpu->valE < cpu->valA || cpu->valE < cpu->valB;
            else if (cpu->op == Sub)
                cpu->flags.of = cpu->valE > cpu->valA || cpu->valE > cpu->valB;

            if (cpu->valE == 0)
                cpu->flags.zf = true;
            if (cpu->valE & ((u64)1 << 63))
                cpu->flags.sf = true;
            break;
        case J:
            sf = cpu->flags.sf;
            zf = cpu->flags.zf;
            of = cpu->flags.of;
            switch (cpu->condition) {
                case None:
                    cpu->cnd = true;
                    break;
                case Le:
                    cpu->cnd = sf || zf;
                    break;
                case L:
                    cpu->cnd = sf && zf;
                    break;
                case E:
                    cpu->cnd = zf;
                    break;
                case Ne:
                    cpu->cnd = !zf;
                    break;
                case Ge:
                    cpu->cnd = !sf || zf;
                    break;
                case G:
                    cpu->cnd = !sf && !zf;
                    break;
            }
            break;
        case CALL:
            cpu->valE = cpu->valB - 8;
            break;
        case RET:
            cpu->valE = cpu->valB + 8;
            break;
        case PUSHQ:
            cpu->valE = cpu->valB - 8;
            break;
        case POPQ:
            cpu->valE = cpu->valB + 8;
            break;
        default:
            break;
    }
}

void memory(CPU* cpu) {
    switch (cpu->curr) {
        case RMMOVQ:
            *((u64*)(cpu->memory + cpu->valE)) = cpu->valA;
            break;
        case MRMOVQ:
            cpu->valM = *((u64*)(cpu->memory + cpu->valE));
            break;
        case CALL:
            *((u64*)(cpu->memory + cpu->valE)) = cpu->valA;
            break;
        case RET:
            cpu->valM = *((u64*)(cpu->memory + cpu->valA));
            break;
        case PUSHQ:
            *((u64*)(cpu->memory + cpu->valE)) = cpu->valA;
            break;
        case POPQ:
            cpu->valM = *((u64*)(cpu->memory + cpu->valA));
            break;
        default:
            break;
    }
}

void writeback(CPU* cpu) {
    switch (cpu->curr) {
        case CMOV:
            if (cpu->cnd)
                set_reg(&cpu->registers, cpu->rB, cpu->valE);
            break;
        case IRMOVQ:
            set_reg(&cpu->registers, cpu->rB, cpu->valE);
            break;
        case MRMOVQ:
            set_reg(&cpu->registers, cpu->rA, cpu->valM);
            break;
        case OPQ:
            set_reg(&cpu->registers, cpu->rB, cpu->valE);
            break;
        case CALL:
            set_reg(&cpu->registers, RSP, cpu->valE);
            break;
        case RET:
            set_reg(&cpu->registers, RSP, cpu->valE);
            break;
        case PUSHQ:
            set_reg(&cpu->registers, RSP, cpu->valE);
            break;
        case POPQ:
            set_reg(&cpu->registers, RSP, cpu->valE);
            set_reg(&cpu->registers, cpu->rA, cpu->valM);
            break;
        default:
            break;
    }
}

void program_counter(CPU* cpu) {
    switch (cpu->curr) {
        case HALT:
            cpu->program_counter = 0;
            break;
        case NOOP:
            cpu->program_counter = cpu->valP;
            break;
        case CMOV:
            cpu->program_counter = cpu->valP;
            break;
        case IRMOVQ:
            cpu->program_counter = cpu->valP;
            break;
        case RMMOVQ:
            cpu->program_counter = cpu->valP;
            break;
        case MRMOVQ:
            cpu->program_counter = cpu->valP;
            break;
        case OPQ:
            cpu->program_counter = cpu->valP;
            break;
        case J:
            if (cpu->cnd)
                cpu->program_counter = cpu->valC;
            else
                cpu->program_counter = cpu->valP;
            break;
        case CALL:
            cpu->program_counter = cpu->valC;
            break;
        case RET:
            cpu->program_counter = cpu->valM;
            break;
        case PUSHQ:
            cpu->program_counter = cpu->valP;
            break;
        case POPQ:
            cpu->program_counter = cpu->valP;
            break;
        default:
            break;
    }
}

int main() {
    CPU cpu = default_cpu();
    while (true) {
        fetch(&cpu);
        decode(&cpu);
        execute(&cpu);
        memory(&cpu);
        writeback(&cpu);
        program_counter(&cpu);
        if (!cpu.stat)
            break;
    }
}
